backend/rohan/src/test/kotlin/ch/chaosconnect/rohan/services/Connect4EngineKtTest.kt

```diff
@@ -0,0 +1,208 @@
+        placedPiece: PieceWithCoordinate,
+        gameBoard: List<PieceWithCoordinate>
+    ) {
+        val board = (gameBoard + placedPiece).asGameBoard()
+        val winningPieces = getWinningPieces(
+            gameBoard = board,
+            placedColumn = placedPiece.column,
+            placedRow = placedPiece.row
+        )
+        assertTrue(winningPieces.isEmpty()) { "Expected no winning pieces, got ${winningPieces.toList()}" }
+    }
+
+    @ParameterizedTest
+    @MethodSource("winningMoves")
+    fun `does identify all pieces if at least 4 are connected`(
+        placedPiece: PieceWithCoordinate,
+        gameBoard: List<PieceWithCoordinate>,
+        expectedPieces: List<PieceWithCoordinate>
+    ) {
+        val board = (gameBoard + placedPiece).asGameBoard()
+        val winningPieces = getWinningPieces(
+            gameBoard = board,
+            placedColumn = placedPiece.column,
+            placedRow = placedPiece.row
+        )
+        assertEquals(
+            expectedPieces.sortedWith(compareBy({ it.column }, { it.row })),
+            winningPieces.sortedWith(compareBy({ it.column }, { it.row }))
+        )
+    }
+
+    companion object {
+        @JvmStatic
+        private fun winningMoves() =
+            Stream.of(
+                Arguments.of(
+                    mockPiece(3, 0, Faction.RED),
+                    horizontalScoreBoard,
+                    (1..6).map { mockPiece(it, 0, Faction.RED) }
+                ),
+                Arguments.of(
+                    mockPiece(7, 0, Faction.RED),
+                    horizontalScoreBoard,
+                    (4..7).map { mockPiece(it, 0, Faction.RED) }
+                ),
+                Arguments.of(
+                    mockPiece(0, 3, Faction.RED),
+                    verticalScoreBoard,
+                    (0..3).map { mockPiece(0, it, Faction.RED) }
+                ),
+                Arguments.of(
+                    mockPiece(0, 3, Faction.YELLOW),
+                    diagonalScoreBoard,
+                    listOf(
+                        mockPiece(0, 3, Faction.YELLOW),
+                        mockPiece(0, 2, Faction.YELLOW),
+                        mockPiece(0, 1, Faction.YELLOW),
+                        mockPiece(0, 0, Faction.YELLOW),
+                        mockPiece(1, 2, Faction.YELLOW),
+                        mockPiece(2, 1, Faction.YELLOW),
+                        mockPiece(3, 0, Faction.YELLOW),
+                    )
+                ),
+                Arguments.of(
+                    mockPiece(5, 3, Faction.RED),
+                    diagonalScoreBoard,
+                    listOf(
+                        mockPiece(2, 0, Faction.RED),
+                        mockPiece(3, 1, Faction.RED),
+                        mockPiece(4, 2, Faction.RED),
+                        mockPiece(5, 3, Faction.RED),
+                    )
+                ),
+            )
+
+        @JvmStatic
+        private fun losingMoves() =
```

Comment:
Just a detail: Some moves here (if not all) don't seem to be losing moves, but rather non-winning moves. Take the moves on `verticalScoreBoard` for example.

Reply:
Losing as in not winning, but you're technically correct 😉